app/lib/dartdoc/backend.dart

```diff
@@ -43,6 +46,60 @@ class DartdocBackend {
 
   DartdocBackend(this._db, this._storage);
 
+  /// Whether the storage bucket has a useable extracted data file.
+  /// Only the existence of the file is checked.
+  // TODO: decide whether we should re-generate the file after a certain age
+  Future<bool> hasValidDartSdkDartdocData() async {
+    final objectName =
+        storage_path.dartSdkDartdocDataName(shared_versions.runtimeVersion);
+    try {
+      final info = await _storage.info(objectName);
+      return info != null;
+    } catch (_) {
+      return false;
+    }
+  }
+
+  /// Upload the generated dartdoc data file for the Dart SDK to the storage bucket.
+  Future uploadDartSdkDartdocData(File file) async {
+    final objectName =
+        storage_path.dartSdkDartdocDataName(shared_versions.runtimeVersion);
+    try {
+      final contentBytes = await file.readAsBytes();
+      await _storage.writeBytes(objectName, _gzip.encode(contentBytes));
+    } catch (e, st) {
+      _logger.warning(
+          'Unable to upload SDK pub dartdoc data file: $objectName', e, st);
+    }
+  }
+
+  /// Read the generated dartdoc data file for the Dart SDK.
+  Future<PubDartdocData> getDartSdkDartdocData() async {
+    final objectName =
+        storage_path.dartSdkDartdocDataName(shared_versions.runtimeVersion);
+    Future<PubDartdocData> load() async {
+      final Map<String, dynamic> map = await _storage
+          .read(objectName)
+          .transform(_gzip.decoder)
+          .transform(utf8.decoder)
+          .transform(json.decoder)
+          .single;
+      return new PubDartdocData.fromJson(map);
+    }
+
+    for (int i = 0; i < 3; i++) {
+      try {
+        return await load();
+      } catch (e, st) {
+        final message =
+            'Unable to read SDK pub dartdoc data file (attempt #${i + 1}): $objectName ';
+        _logger.info(message, e, st);
```

Comment:
Woah. This seems flaky. What's the problem here?

Reply:
IIRC it was about the time where we have experienced several timeouts and 500s from the dartdoc's storage bucket both in prod and in staging. We can be more optimistic now, shall I remove?

Reply:
If you think it's stable? If we try once here, and it fails, is it later retried at a higher level?

Reply:
Yes, it is retried at a higher level, removing the redundancy.